node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,796 @@
+func ParseConfig(ibcConfig string) (ConnectionConfig, string, error) {
+	connections := make([]ConnectionConfigEntry, 0)
+	features := ""
+
+	if ibcConfig == "" {
+		// This is not an error if IBC is not enabled.
+		return connections, features, nil
+	}
+
+	// The config string is json formatted like this: `[{"ChainID":18,"ConnID":"connection-0"},{"ChainID":19,"ConnID":"connection-1"}]`
+	err := json.Unmarshal([]byte(ibcConfig), &connections)
+	if err != nil {
+		return connections, features, fmt.Errorf("failed to parse IBC config string: %s, error: %w", ibcConfig, err)
+	}
+
+	// Build the feature string.
+	for _, ch := range connections {
+		if features == "" {
+			features = "ibc:"
+		} else {
+			features += ","
+		}
+		features += fmt.Sprintf("%s:%s", ch.ChainID.String(), ch.ConnID)
+	}
+
+	return connections, features, nil
+}
+
+// NewWatcher creates a new IBC contract watcher
+func NewWatcher(
+	wsUrl string,
+	lcdUrl string,
+	contractAddress string,
+	ConnectionConfig ConnectionConfig,
+	channelData ChannelData,
+) *Watcher {
+	return &Watcher{
+		wsUrl:            wsUrl,
+		lcdUrl:           lcdUrl,
+		contractAddress:  contractAddress,
+		connectionConfig: ConnectionConfig,
+		channelData:      channelData,
+		connectionIdMap:  make(map[string]string),
+	}
+}
+
+// ConvertUrlToTendermint takes a URL and does the following conversions if necessary:
+// - Converts "ws://" to "http:".
+// - Strips "/websocket" off the end.
+func ConvertUrlToTendermint(input string) (string, error) {
+	input = strings.TrimPrefix(input, "ws://")
+	input = strings.TrimPrefix(input, "http://")
+	input = strings.TrimSuffix(input, "/websocket")
+	return "http://" + input, nil
+}
+
+// clientRequest is used to subscribe for events from the contract.
+type clientRequest struct {
+	JSONRPC string `json:"jsonrpc"`
+	// A String containing the name of the method to be invoked.
+	Method string `json:"method"`
+	// Object to pass as request parameter to the method.
+	Params [1]string `json:"params"`
+	// The request id. This can be of any type. It is used to match the
+	// response with the request that it is replying to.
+	ID uint64 `json:"id"`
+}
+
+const (
+	// The IBC receiver contract publishes wasm events, not execute events.
+	contractAddressFilterKey = "wasm._contract_address"
+	contractAddressLogKey    = "_contract_address"
+)
+
+// Run is the runnable for monitoring the IBC contract on wormchain.
+func (w *Watcher) Run(ctx context.Context) error {
+	w.logger = supervisor.Logger(ctx)
+
+	errC := make(chan error)
+	defer close(errC)
+
+	// Rebuild these from scratch every time the watcher restarts.
+	w.connectionMap = make(map[string]*connectionEntry)
+	w.chainMap = make(map[vaa.ChainID]*connectionEntry)
+
+	// Build our internal data structures based on the config passed in.
+	for idx, chainToMonitor := range w.connectionConfig {
```

Comment:
There is error checking here. If it were done in the constructor, it would require a panic, which I'm not crazy about. 

I changed it so that this initialization will only be done the first time Run() is invoked.